<a href="https://colab.research.google.com/github/Aiswarya96-creator/Aiswarya96-creator/blob/main/data_aquisition_case_study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
#import requests for only if url's are take from internet.if the files are already given then dont use
import sqlite3
import pandas as pd

data extraction

In [2]:
population=pd.read_csv("/content/population.csv")
population

,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525
5,United Kingdom,67886011
6,France,65273511
7,Italy,60461826
8,Canada,38005238
9,Australia,25687041


In [3]:
population.head()


,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525


In [4]:
population.shape

(25, 2)

In [5]:
population.dtypes

,0
country,object
population,int64


In [6]:
gdp=pd.read_excel("/content/gdp.xlsx")
gdp

,Country,GDP
0,United States,30500000
1,China,19230000
2,India,4187000
3,Japan,4180000
4,Germany,4744000
5,United Kingdom,3839000
6,France,3211000
7,Italy,2420000
8,Canada,2250000
9,Australia,1880000


In [7]:
gdp.columns

Index(['Country', 'GDP'], dtype='object')

In [8]:
gdp.dtypes

,0
Country,object
GDP,int64


In [9]:
users=pd.read_json("/content/internet_users.json")
users

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000
5,United Kingdom,66000000
6,France,62000000
7,Italy,56000000
8,Canada,36000000
9,Australia,24000000


In [10]:
pd.DataFrame(users)

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000
5,United Kingdom,66000000
6,France,62000000
7,Italy,56000000
8,Canada,36000000
9,Australia,24000000


In [11]:
litracy=pd.read_xml("/content/literacy_rate.xml")
litracy

,name,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


In [12]:
pd.DataFrame(litracy)

,name,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


data understanding

In [13]:
population.shape[0]



25

In [14]:
population.shape[1]

2

In [15]:
population.columns

Index(['country', 'population'], dtype='object')

In [16]:
population.dtypes

,0
country,object
population,int64


In [17]:
population.isna().sum()

,0
country,0
population,0


In [18]:
population.duplicated().sum()

np.int64(0)

In [19]:
litracy.shape[0]

25

In [20]:
litracy.shape[1]

2

In [21]:
litracy.columns

Index(['name', 'literacy_rate'], dtype='object')

In [22]:
litracy.dtypes

,0
name,object
literacy_rate,float64


In [23]:
litracy.isna().sum()

,0
name,0
literacy_rate,0


In [24]:
litracy.duplicated().sum()

np.int64(0)

In [25]:
users.shape[0]

25

In [26]:
users.shape[1]

2

In [27]:
users.columns

Index(['country', 'internet_users'], dtype='object')

In [28]:
users.dtypes

,0
country,object
internet_users,int64


In [29]:
users.isna().sum()

,0
country,0
internet_users,0


In [30]:
users.duplicated().sum()

np.int64(0)

In [31]:
gdp.shape[0]

25

In [32]:
gdp.shape[1]

2

In [33]:
gdp.columns

Index(['Country', 'GDP'], dtype='object')

In [34]:
gdp.dtypes

,0
Country,object
GDP,int64


In [35]:
gdp.isna().sum()

,0
Country,0
GDP,0


In [36]:
gdp.duplicated().sum()

np.int64(0)

data transformation

In [37]:
gdp.rename(columns={'Country':'country'},inplace=True)
gdp

,country,GDP
0,United States,30500000
1,China,19230000
2,India,4187000
3,Japan,4180000
4,Germany,4744000
5,United Kingdom,3839000
6,France,3211000
7,Italy,2420000
8,Canada,2250000
9,Australia,1880000


In [38]:
litracy.rename(columns={'name':'country'},inplace=True)
litracy

,country,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


In [39]:
litracy['literacy_rate']=pd.to_numeric(litracy['literacy_rate'])
litracy


,country,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0


In [40]:
population['population']=pd.to_numeric(population['population'])
population

,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525
5,United Kingdom,67886011
6,France,65273511
7,Italy,60461826
8,Canada,38005238
9,Australia,25687041


In [41]:
users['internet_users']=pd.to_numeric(users['internet_users'])
users

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000
5,United Kingdom,66000000
6,France,62000000
7,Italy,56000000
8,Canada,36000000
9,Australia,24000000


In [42]:
gdp['GDP']=pd.to_numeric(gdp['GDP'])
gdp

,country,GDP
0,United States,30500000
1,China,19230000
2,India,4187000
3,Japan,4180000
4,Germany,4744000
5,United Kingdom,3839000
6,France,3211000
7,Italy,2420000
8,Canada,2250000
9,Australia,1880000


In [43]:
merged=population.merge(gdp,on="country")
merged=merged.merge(users,on="country")
merged=merged.merge(litracy,on="country")
merged

,country,population,GDP,internet_users,literacy_rate
0,United States,331002651,30500000,312000000,99.0
1,China,1411778724,19230000,1080000000,97.2
2,India,1380004385,4187000,850000000,77.7
3,Japan,125836021,4180000,118000000,99.0
4,Germany,83240525,4744000,79000000,99.0
5,United Kingdom,67886011,3839000,66000000,99.0
6,France,65273511,3211000,62000000,99.0
7,Italy,60461826,2420000,56000000,99.0
8,Canada,38005238,2250000,36000000,99.0
9,Australia,25687041,1880000,24000000,99.0


In [46]:
merged["Internet Penetration (%)"] = (
    merged["internet_users"] /
    merged["population"] *100
).round(2)

merged["GDP Per Capita"] = (
    merged["GDP"] /
    merged["population"]
).round(2)

merged.head()

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06


load data

In [48]:
conn = sqlite3.connect("country_data.db")

merged.to_sql(
    "country_statistics",
    conn,
    if_exists="replace",
    index=False
)

25

In [49]:
query = """
SELECT *
FROM country_statistics
LIMIT 5;
"""

pd.read_sql(query,conn)

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
2,India,1380004385,4187000,850000000,77.7,61.59,0.00
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06


In [50]:
merged.nlargest(5,"GDP")[["country","GDP"]]

,country,GDP
0,United States,30500000
1,China,19230000
4,Germany,4744000
2,India,4187000
3,Japan,4180000


In [51]:
merged.nlargest(5,"population")[["country","population"]]

,country,population
1,China,1411778724
2,India,1380004385
0,United States,331002651
15,Indonesia,273523615
10,Brazil,212559417


In [52]:
merged["literacy_rate"].mean()

np.float64(97.248)

In [53]:
merged[merged["literacy_rate"]>90]

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.06
6,France,65273511,3211000,62000000,99.0,94.98,0.05
7,Italy,60461826,2420000,56000000,99.0,92.62,0.04
8,Canada,38005238,2250000,36000000,99.0,94.72,0.06
9,Australia,25687041,1880000,24000000,99.0,93.43,0.07
10,Brazil,212559417,2330000,181000000,93.2,85.15,0.01


In [54]:
merged[
merged["Internet Penetration (%)"]>70
]

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
1,China,1411778724,19230000,1080000000,97.2,76.50,0.01
3,Japan,125836021,4180000,118000000,99.0,93.77,0.03
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.06
6,France,65273511,3211000,62000000,99.0,94.98,0.05
7,Italy,60461826,2420000,56000000,99.0,92.62,0.04
8,Canada,38005238,2250000,36000000,99.0,94.72,0.06
9,Australia,25687041,1880000,24000000,99.0,93.43,0.07
10,Brazil,212559417,2330000,181000000,93.2,85.15,0.01


sql

In [55]:
merged.nlargest(5,"GDP Per Capita")[
["country","GDP Per Capita"]
]

,country,GDP Per Capita
21,Switzerland,0.11
23,Norway,0.11
0,United States,0.09
24,Singapore,0.09
9,Australia,0.07


In [56]:
merged["population"].sum()

np.int64(4720039725)

In [57]:
merged["Internet Penetration (%)"].mean()

np.float64(89.6596)

In [58]:
merged.nlargest(
1,
"internet_users"
)[["country","internet_users"]]

,country,internet_users
1,China,1080000000


In [59]:
merged["literacy_rate"].corr(
merged["Internet Penetration (%)"]
)

np.float64(0.77279756813985)

In [66]:
query = """
SELECT country, GDP
FROM country_statistics
ORDER BY GDP DESC
LIMIT 5;
"""
pd.read_sql(query,conn)

,country,GDP
0,United States,30500000
1,China,19230000
2,Germany,4744000
3,India,4187000
4,Japan,4180000


In [67]:
query = """
SELECT AVG(literacy_rate)
FROM country_statistics;
"""
pd.read_sql(query,conn)

,AVG(literacy_rate)
0,97.248


In [68]:
query = """
SELECT country,literacy_rate
FROM country_statistics
WHERE literacy_rate>90;
"""
pd.read_sql(query,conn)

,country,literacy_rate
0,United States,99.0
1,China,97.2
2,Japan,99.0
3,Germany,99.0
4,United Kingdom,99.0
5,France,99.0
6,Italy,99.0
7,Canada,99.0
8,Australia,99.0
9,Brazil,93.2


In [69]:
query = """
SELECT country, "Internet Penetration (%)"
FROM country_statistics
ORDER BY "Internet Penetration (%)" DESC
LIMIT 5;
"""
pd.read_sql(query,conn)

,country,Internet Penetration (%)
0,Switzerland,98.21
1,Norway,97.76
2,Saudi Arabia,97.66
3,Sweden,97.55
4,Netherlands,97.47


In [70]:
query = """
SELECT SUM(population)
FROM country_statistics;
"""
pd.read_sql(query, conn)

,SUM(population)
0,4720039725


In [71]:
query = """
SELECT country,
       "GDP Per Capita"
FROM country_statistics
ORDER BY "GDP Per Capita" DESC
LIMIT 1;
"""
pd.read_sql(query, conn)

,country,GDP Per Capita
0,Switzerland,0.11


In [72]:
merged.to_csv(
    "country_statistics.csv",
    index=False
)
pd.read_sql(query, conn)

,country,GDP Per Capita
0,Switzerland,0.11


In [73]:
merged.sort_values(
    by="GDP Per Capita",
    ascending=False
).head(10)

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
21,Switzerland,8654622,940000,8500000,99.0,98.21,0.11
23,Norway,5421241,580000,5300000,100.0,97.76,0.11
24,Singapore,5850342,530000,5600000,97.5,95.72,0.09
0,United States,331002651,30500000,312000000,99.0,94.26,0.09
9,Australia,25687041,1880000,24000000,99.0,93.43,0.07
20,Netherlands,17441139,1150000,17000000,99.0,97.47,0.07
4,Germany,83240525,4744000,79000000,99.0,94.91,0.06
8,Canada,38005238,2250000,36000000,99.0,94.72,0.06
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.06
22,Sweden,10353442,650000,10100000,99.0,97.55,0.06


ETL process explaination
data pipeline that ingested information from CSV, Excel, JSON, and XML files, cleaned and standardized the data using Python's Pandas library, integrated the datasets by country, and calculated new metrics for Internet Penetration and GDP Per Capita before storing the final result in a SQLite database. Subsequent analysis revealed that the United States holds the highest GDP, China leads in both population and internet user count, the average literacy rate across the studied countries is 97.248%, and there is a strong positive correlation of approximately 0.773 between literacy rates and internet penetration.